Add support of ARC MWDT libc to Zephyr

evgeniy-paltsev opened the pull request:
ARC MWDT toolchain doesn't provide NEWLIB libc. Let's add support of of ARC MWDT libc to Zephyr.

This is prerequisite of adding ARC MWDT C++ library support to Zephyr.

evgeniy-paltsev committed:
tests: posix: common: add option to build with arcmwdt libc

Add testcase to tests/posix/common with ARC MWDT libc usage.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: LIB: add ARC MWDT libc support

ARC MWDT toolchain doesn't provide newlib, let's add support
of ARC MWDT libc to Zephyr

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: LIB: MWDT: add stdout hooks, timespec header

ARC MWDT toolchain misses stdout hooks implementation and
itimerspec structure in timespec header. Let's add them in
arcmwdt compatibility layer.

The implementation was inspired by libc-hooks.c for NEWLIB.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: MWDT: workaround for double clockid_t definition

Don't define clockid_t in Zephyr in case of ARC MWDT libc
is used as ARC MWDT libc provides clockid_t.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: MWDT: workaround missing ssize_t definition

ARC MWDT don't provide ssize_t type which is used in Zephyr.
Define ssize_t for ARC MWDT toolchain.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: MWDT: workaround paddr_t defined in both Zephyr and toolchain

MWDT provides paddr_t type and it conflicts witn Zephyr definition:
- Zephyr defines paddr_t as a uintptr_t
- MWDT defines paddr_t as a unsigned long
This causes extra warnings. However we can safely define
paddr_t as a unsigned long for the case when MWDT toolchain is used as
they are both unsighned, have same size and aligning.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: MWDT: workaround missing file IO related defines

ARC MWDT libc misses some file IO related defines, let's
add them in ARC MWDT libc compatibility layer.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: MWDT: LIB: implement ___errno hook

Implement ___errno so ARC MWDT libc can use proper (Zephyr) errno

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: MWDT: don't link MetaWare hostlink lib ever

MetaWare hostlink is a library of system calls for debugging.
We don't use in case of Zephyr, however we provide real system
calls implementations. Let's pass -Hhostlib= to MWDT linker to
eliminate the MetaWare hostlink library from linkage.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

abrodkin commented:
Needs to be re-based on top of https://github.com/zephyrproject-rtos/zephyr/pull/36744 once it is merged.

ruuddw commented on lib/libc/arcmwdt/arcmwdt-string.c:
```diff
@@ -0,0 +1,30 @@
+ * @return number of bytes in fixed-size string <s>
+ */
+
+size_t strnlen(const char *s, size_t maxlen)
```
No strnlen in MW libc? maybe file a enhancement request for that...
I do see strnlen_s is part of MW, use that instead?

evgeniy-paltsev replied:
Fixed. Now we simply call strnlen_s in strnlen.

ruuddw commented on lib/libc/arcmwdt/include/fcntl.h:
```diff
@@ -0,0 +1,21 @@
+#ifndef ZEPHYR_LIB_LIBC_ARCMWDT_INCLUDE_FCNTL_H_
+#define ZEPHYR_LIB_LIBC_ARCMWDT_INCLUDE_FCNTL_H_
+
+#define O_CREAT    0x0200
```
Aren't these define in MW sys/fctl?
Can you simply define them here, without ensuring that the MW implementation isn't using other constants?

evgeniy-paltsev replied:
Fixed. Now we only define O_NONBLOCK which is missing in MWDT fcntl.h header. We also check that our O_NONBLOCK value doesn't conflict with any other file IO defines which came from MWDT fcntl.h header.

ruuddw replied:
Nice solution, thanks.

ruuddw reviewed (approved):
Ready for merge imho.

evgeniy-paltsev commented:
The warnings in compliance check (`WARNING:NEW_TYPEDEFS: do not add new typedefs`) are related to adding types missing in MWDT toolchain, so they should be ignored.

evgeniy-paltsev commented:
This PR has enough approvals and it was kept open enough time so everyone could share their concerns (if any). I guess we can safely merge it now.

stephanosio commented on lib/libc/arcmwdt/libc-hooks.c:
```diff
@@ -0,0 +1,113 @@
+	_stdout_hook = hook;
+}
+
+int z_impl_zephyr_fputc(int c, FILE *stream)
```
Here you are implicitly implementing the "minimal libc" hook interface functions:
https://github.com/zephyrproject-rtos/zephyr/blob/775cf55d9268201be6f42523c80ffad808fd3209/include/sys/libc-hooks.h#L19-L37

You should add `#elif defined(CONFIG_ARCMWDT_LIBC)` there and add only the required hook functions for the ARC MWDT libc.

evgeniy-paltsev replied:
Fixed in new implementation, thanks!

stephanosio reviewed (changes requested):
The `libc-hooks.c` implementation does not look sane to me.

I assume `fputs` and others are already implemented by the ARC MWDT libc, but they are declared weak and therefore overridable.

In that case, only the most fundamental (interface) functions should be overridden by the `libc-hooks.c` (e.g. `write()` and `read()` as with the newlib implementation). I highly doubt the ARC MWDT libc implementation just happens to have the same interface definition as what we currently have in `stdout_console.c` of the minimal libc.

evgeniy-paltsev commented on lib/libc/arcmwdt/libc-hooks.c:
```diff
@@ -0,0 +1,113 @@
+#include <syscalls/zephyr_fputc_mrsh.c>
+#endif
+
+int fputc(int c, FILE *stream)
```
Yep, you are correct. I've dropped it in a new implementation.

evgeniy-paltsev commented on include/toolchain/mwdt.h:
```diff
@@ -61,6 +61,17 @@
+	#define _MLIBC_RESTRICT restrict
+#endif
+
+#endif /* _MLIBC_RESTRICT_defined */
```
Removed.

evgeniy-paltsev commented on lib/libc/arcmwdt/libc-hooks.c:
```diff
@@ -0,0 +1,113 @@
+/* stdout_console.c */
```
Fixed.

stephanosio reviewed (changes requested):
Looks ok overall. Some minor fixes.

evgeniy-paltsev committed:
ARC: MWDT: LIB: implement _istty hook

Implement _istty hook as it is required for proper setup of
STDIN/STDOUT/STDERR buffering.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

stephanosio commented on lib/libc/arcmwdt/libc-hooks.c:
```diff
@@ -0,0 +1,77 @@
+static inline int z_vrfy_z_zephyr_write_stdout(const void *buf, int nbytes)
+{
+	Z_OOPS(Z_SYSCALL_MEMORY_READ(buf, nbytes));
+	return z_impl_zephyr_write_stdout((const void *)buf, nbytes);
```
Cast should not be necessary.
```suggestion
	return z_impl_zephyr_write_stdout(buf, nbytes);
```

evgeniy-paltsev replied:
Thanks for pointting, fixed!

stephanosio commented on lib/libc/arcmwdt/libc-hooks.c:
```diff
@@ -0,0 +1,77 @@
+ * It's require to implement _isatty to have STDIN/STDOUT/STDERR buffered
+ * properly.
+ */
+int _isatty(int file) {
+	if (file == STDIN_FILENO || file == STDOUT_FILENO || file == STDERR_FILENO) {
```
```
-:308: ERROR:OPEN_BRACE: open brace '{' following function definitions go on the next line
#308: FILE: lib/libc/arcmwdt/libc-hooks.c:66:
+int _isatty(int file) {
```

```suggestion
int _isatty(int file)
{
	if (file == STDIN_FILENO || file == STDOUT_FILENO || file == STDERR_FILENO) {
```

evgeniy-paltsev committed:
ARC: MWDT: link mwdt libc only if it is required

We don't need MWDT libc if we are using minimal one provided by
Zephyr. So let's avoid linking mwdt libc if CONFIG_MINIMAL_LIBC
is enabled.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: MWDT: generate access to builtin funcs only if ARCMWDT_LIBC is used

The MWDT compiler can replace some code with call to builtin
functions. We can't rely on these functions presence if we
don't use MWDT libc.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

ruuddw reviewed (approved):
re-approved latest version

cfriedt commented:
Thanks @evgeniy-paltsev - will merge as soon as it's green. And thanks to everyone for the reviews ;-)

evgeniy-paltsev commented:
@cfriedt ok, now there are only a few warnings about adding typedefs missing in MWDT toolchain, so they should be ignored.

Thanks for pointing to the issue with too long commit line!
